internal/service/lakeformation/resource_lf_tags.go

```diff
@@ -0,0 +1,518 @@
+							Required: true,
+						},
+						"wildcard": {
+							Type:     schema.TypeBool,
+							Default:  false,
+							ForceNew: true,
+							Optional: true,
+							AtLeastOneOf: []string{
+								"table_with_columns.0.column_names",
+								"table_with_columns.0.wildcard",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceResourceLFTagsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).LakeFormationConn
+
+	input := &lakeformation.AddLFTagsToResourceInput{
+		Resource: &lakeformation.Resource{},
+	}
+
+	if v, ok := d.GetOk("catalog_id"); ok {
+		input.CatalogId = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("database"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
+		input.Resource.Database = ExpandDatabaseResource(v.([]interface{})[0].(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("lf_tag"); ok && v.(*schema.Set).Len() > 0 {
+		input.LFTags = expandLFTagPairs(v.(*schema.Set).List())
+	}
+
+	if v, ok := d.GetOk("table"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
+		input.Resource.Table = ExpandTableResource(v.([]interface{})[0].(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("table_with_columns"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
+		input.Resource.TableWithColumns = expandTableColumnsResource(v.([]interface{})[0].(map[string]interface{}))
+	}
+
+	var output *lakeformation.AddLFTagsToResourceOutput
+	err := resource.Retry(IAMPropagationTimeout, func() *resource.RetryError {
+		var err error
+		output, err = conn.AddLFTagsToResource(input)
+		if err != nil {
+			if tfawserr.ErrCodeEquals(err, lakeformation.ErrCodeConcurrentModificationException) {
+				return resource.RetryableError(err)
+			}
+			if tfawserr.ErrMessageContains(err, "AccessDeniedException", "is not authorized") {
+				return resource.RetryableError(err)
+			}
+
+			return resource.NonRetryableError(fmt.Errorf("error creating Lake Formation LF Tag Resource: %w", err))
+		}
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		output, err = conn.AddLFTagsToResource(input)
+	}
+
+	if err != nil {
+		return names.DiagError(names.LakeFormation, names.ErrActionCreating, resourceLFTags, input.String(), err)
+	}
+
+	if output != nil && len(output.Failures) > 0 {
+		var failures *multierror.Error
+
+		for _, v := range output.Failures {
+			if v.LFTag == nil || v.Error == nil {
+				continue
+			}
+
+			origErr := fmt.Errorf("catalog id:%s, tag key:%s, values:%+v", aws.StringValue(v.LFTag.CatalogId), aws.StringValue(v.LFTag.TagKey), aws.StringValueSlice(v.LFTag.TagValues))
+			err := awserr.New(aws.StringValue(v.Error.ErrorCode), aws.StringValue(v.Error.ErrorMessage), origErr)
+			failures = multierror.Append(failures, err)
+		}
+
+		if failures.Len() > 0 {
+			return names.DiagError(names.LakeFormation, names.ErrActionCreating, resourceLFTags, "", failures.ErrorOrNil())
```

Comment:
should we be looking to update the terraform state for all/any of the successful lf_tag applies before exiting with error if there are any failures?

Reply:
We could return a `diag.Warning` for each `LFTag` that fails and only an overall `diag.Error` if `failures.Len() == len(input. LFTags)`?

Reply:
The trouble is that we don't update state in Create. I'm not sure how helpful it would be in that these are very fast operations and on the next successful Read, the state will be fixed. I think we need to stick with the framework and let Read handle state updates unless there's some really compelling reason.

Reply:
I don't know where else we're doing something like this so it might be a first. Stack of failure warning, as requested. In my testing, I haven't seen exactly what a failure _is_. For example, if you send duplicates of a key, you get an error back on the whole `AddLFTagsToResource` operation so this logic is never reached. Hopefully it won't confuse people.